locker.go

```diff
@@ -0,0 +1,344 @@
+
+	for _, key := range keys {
+		var lockCtx context.Context
+		var lockCancel context.CancelFunc
+
+		if lockCtx, lockCancel, err = l.tryLock(sessionCtx, key); lockCtx != nil {
+			return lockCtx, func() { lockCancel(); sessionCancel() }
+		}
+	}
+
+	sessionCancel()
+	return errorContext(ctx, coalesceError(err, Unlocked))
+}
+
+func (l *Locker) tryLock(ctx context.Context, key string) (context.Context, context.CancelFunc, error) {
+	client := l.client()
+	session := contextSession(ctx)
+
+	tryLockCtx, tryLockCancel := context.WithTimeout(ctx, l.lockDelay())
+	defer tryLockCancel()
+
+	locked, err := client.acquireLock(tryLockCtx, key, string(session.ID))
+	if !locked || err != nil {
+		return nil, nil, err
+	}
+
+	lock := newLockCtx(ctx, key, client)
+	return lock, lock.cancel, nil
+}
+
+func (l *Locker) withSession(ctx context.Context, name string, args ...interface{}) (context.Context, context.CancelFunc) {
+	if ctx.Value(SessionKey) != nil {
+		// The context that was used to create the lock is already a session, we
+		// can use it directly instead of recreating one.
+		return ctx, func() {}
+	}
+	lockDelay := l.lockDelay()
+	return WithSession(ctx, Session{
+		Client:    l.client(),
+		Name:      fmt.Sprintf(name, args...),
+		Behavior:  Release,
+		LockDelay: lockDelay,
+		TTL:       lockDelay * 2,
+	})
+}
+
+func (l *Locker) client() *Client {
+	if client := l.Client; client != nil {
+		return client
+	}
+	return DefaultClient
+}
+
+func (l *Locker) lockDelay() time.Duration {
+	if delay := l.LockDelay; delay != 0 {
+		return delay
+	}
+	return 15 * time.Second
+}
+
+// Lock calls DefaultLocker.Lock.
+func Lock(ctx context.Context, keys ...string) (context.Context, context.CancelFunc) {
+	return DefaultLocker.Lock(ctx, keys...)
+}
+
+// TryLock calls DefaultLocker.TryLock.
+func TryLock(ctx context.Context, keys ...string) (context.Context, context.CancelFunc) {
+	return DefaultLocker.TryLock(ctx, keys...)
+}
+
+var (
+	// DefaultLocker is the default Locker used by the package-level lock
+	// management functions.
+	DefaultLocker = &Locker{}
+
+	// LocksKey is used to lookup the keys held by a lock from its associated
+	// context.
+	LocksKey = &contextKey{"consul-locks-key"}
+
+	// Unlocked is the error returned by contexts when the lock they were
+	// associated with has been lost.
+	Unlocked = errors.New("unlocked")
+)
+
+func newLockCtx(ctx context.Context, key string, client *Client) *lockCtx {
+	l := &lockCtx{
+		client: client,
+		ctx:    ctx,
+		key:    key,
+		done:   make(chan struct{}),
+	}
+	go l.run()
+	return l
+}
+
+type lockCtx struct {
+	client *Client
+	ctx    context.Context
+	key    string
+	err    atomic.Value
+	once   sync.Once
+	done   chan struct{}
+}
+
+func (l *lockCtx) Deadline() (time.Time, bool) {
+	return l.ctx.Deadline()
+}
+
+func (l *lockCtx) Done() <-chan struct{} {
+	return l.done
+}
+
+func (l *lockCtx) Err() error {
+	err, _ := l.err.Load().(error)
```

Comment:
Because it could be unset, in which case it would panic if we don't use this form.